quixstreams/models/topics/manager.py

```diff
@@ -201,147 +202,144 @@ def register(self, topic: Topic) -> Topic:
-                    raise TopicConfigurationMismatch(
-                        f'changelog topic "{changelog_topic.name}" partition count '
-                        f'does not match its source topic "{source_topic.name}": '
-                        f"expected {source_config.num_partitions}, "
-                        f'got {changelog_config.num_partitions}"'
-                    )
-                if (
-                    changelog_config.replication_factor
-                    != source_config.replication_factor
-                ):
-                    raise TopicConfigurationMismatch(
-                        f'changelog topic "{changelog_topic.name}" replication factor '
-                        f'does not match its source topic "{source_topic.name}": '
-                        f"expected {source_config.replication_factor}, "
-                        f'got {changelog_config.replication_factor}"'
-                    )
-
-        logger.info("Kafka topics validation complete")
+    @classmethod
+    def derive_topic_config(cls, *topics: Topic) -> TopicConfig:
+        """
+        Derive a topic config based on one or more input Topic configs.
+        To be used for generating the internal changelogs and repartition topics.
+
+        This method expects that Topics contain "retention.ms" and "retention.bytes"
+        config values.
+
+        Multiple topics are expected for merged and joins streams.
+        """
+        if not topics:
+            raise ValueError("At least one Topic must be passed")
+
+        # Pick the maximum number of partitions across topics
+        num_partitions = max(
+            t.broker_config.num_partitions
+            for t in topics
+            if t.broker_config.num_partitions is not None
+        )
+
+        # Pick the maximum replication factor
+        replication_factor = max(
+            t.broker_config.replication_factor
+            for t in topics
+            if t.broker_config.replication_factor is not None
+        )
+
+        # Pick the maximum "retention.bytes" value across topics. -1 means infinity
+        retention_bytes_values = [
+            int(t.broker_config.extra_config["retention.bytes"]) for t in topics
+        ]
+        retention_bytes = (
+            -1 if -1 in retention_bytes_values else max(retention_bytes_values)
+        )
+
+        # Pick maximum "retention.ms" value across topics. -1 means infinity
+        retention_ms_values = [
+            int(t.broker_config.extra_config["retention.ms"]) for t in topics
+        ]
+        retention_ms = -1 if -1 in retention_ms_values else max(retention_ms_values)
+
+        return TopicConfig(
+            num_partitions=num_partitions,
+            replication_factor=replication_factor,
+            extra_config={
+                "retention.bytes": str(retention_bytes),
+                "retention.ms": str(retention_ms),
+            },
+        )
+
+    def stream_id_from_topics(self, *topics: Topic) -> str:
```

Comment:
I wonder if there is a reason for packing positional arguments versus accepting a single argument with precise typing.
```suggestion
    def stream_id_from_topics(self, topics: FrozenSet[Topic]) -> str:
```

Reply:
I agree with converting `*topics` to `topics`. This func is always called with the collection.
But I'd make the type to be `Iterable` since it's more than enough for this method (it doesn't really need to know that the set is being passed)